### ISSUE TITLE
Why is the accuracy overestimated?

### ISSUE THREAD

**Issue by mstokes42** [apache/incubator-mxnet#8601]
I have trained a model, and want to see how it performs on both the training and testing data. I'm using a SoftMax output with 2 hidden nodes for a binary classification problem.

```
# predict accuracy
acc = mx.metric.Accuracy()

lenet_model.score(train_iter, acc)
print('Training:',acc,F1)

lenet_model.score(val_iter, acc)
print('Testing:',acc,F1)

output = lenet_model.predict(val_iter)
```
This yields perfect results on the training data, and not very good results on the test data

Training: EvalMetric: {'accuracy': 1.0}
Testing: EvalMetric: {'accuracy': 0.51111111111111107}

However, when I go and look at the individual predictions, I find that this accuracy on the test set is overestimated.

```
y_pred = np.round(lenet_model.predict(train_iter).asnumpy()[:,1])
y_true = phen[:ntrain]
print(metrics.confusion_matrix(y_true, y_pred))

y_pred = np.round(lenet_model.predict(val_iter).asnumpy()[:,1])
y_true = phen[ntrain:]
print(metrics.confusion_matrix(y_true, y_pred))
```
This yields the following confusion matrices:
[[127   0]
 [  0  76]]
[[21 14]
 [13  3]]

I can see that the training set is indeed perfectly classified. The test set, however, only gets 24 of 51 correct, for an accuracy of 47%. Why does the model.score() function tell me that I have achieved 51% accuracy?

**Comment by eric-haibin-lin** [comment]
Metrics are stateful. If you pass a new Acc metric, is the estimation correct?
```
lenet_model.score(val_iter, mx.metric.Accuracy())
```

**Comment by mstokes42** [comment]
I figured out the problem was with my batch_size on the val_iter. The sample size was not a multiple of the batch size, so some samples were being discarded by the model.score call that uses val_iter. By changing the batch size to equal the number of samples, the accuracy comes out correct.

Now I'm having an issue where by model performs well on the training data, but performs worse than random on the test data. I'd be better off taking the output of the classifier and predicting the opposite. When I apply the model to a different held-out validation set, though, the model again performs almost perfectly... what could explain this behavior?

**Comment by eric-haibin-lin** [comment]
Are you using k-fold validation?

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by leleamol** [comment]
Recommended Labels: "Question","Need Triage","Pending Requester Info"

**Comment by szha** [comment]
@leleamol @marcoabreu please refrain from manually adding "Need Triage" label while triaging. Thanks.

**Comment by vishaalkapoor** [comment]
@mstokes42 If your confusion matrix is representative of your test/train set, it looks like you're using very few (~ 100) training examples with what appears to be LeNet-5. This is about 2 orders of magnitude too little data; in [Lecun98](http://vision.stanford.edu/cs598_spring07/papers/Lecun98.pdf) LeNet-5 was trained using MNIST (60k 32x32 training images).

The reason you're getting 100% accuracy on your train set is that LeNet-5 is memorizing your training data and not learning enough to generalize to your test set. 

Take a look at the Convolutional Neural networks section for some examples with training sets:
https://gluon.mxnet.io/chapter04_convolutional-neural-networks/cnn-scratch.html
E.g.
`
batch_size = 64
num_inputs = 784
num_outputs = 10
def transform(data, label):
    return nd.transpose(data.astype(np.float32), (2,0,1))/255, label.astype(np.float32)
train_data = gluon.data.DataLoader(gluon.data.vision.MNIST(train=True, transform=transform),
                                      batch_size, shuffle=True)
test_data = gluon.data.DataLoader(gluon.data.vision.MNIST(train=False, transform=transform),
                                     batch_size, shuffle=False)
`
Additionally, if this is the issue, please feel free to close the issue. Thanks!

**Comment by vandanavk** [comment]
@sandeep-krishnamurthy Please close this issue based on @vishaalkapoor's comments above.

@mstokes42 Please feel free to reopen this issue if it has been closed in error.
